### ISSUE TITLE
[CollectionView] [Windows] memory leaks when ItemsSource has changed

### ISSUE THREAD

**Issue by ivan-todorov-progress** [dotnet/maui#13393]
### Description

It seems that the `CollectionView` is leaking memory every time its `ItemsSource` property has changed.

Due to a lack of a proper memory profiler on all platforms, I am using a poor man's memory tracker implementation that stores a collection of `WeakReference`s and observes their `IsAlive` status:

```csharp
public static class MemoryTracker
{
    private static readonly List<WeakReference> weakReferences = new List<WeakReference>();

    public static readonly BindableProperty IsTrackedProperty =
        BindableProperty.CreateAttached("IsTracked", typeof(bool), typeof(MemoryTracker),
            false, propertyChanged: OnIsTrackedChanged);

    public static int TotalObjectCount => weakReferences.Count;

    public static int AliveObjectCount => weakReferences.Count(weakReference => weakReference.IsAlive);

    public static bool GetIsTracked(BindableObject bindable)
    {
        return (bool)bindable.GetValue(IsTrackedProperty);
    }

    public static void SetIsTracked(BindableObject bindable, bool isTracked)
    {
        bindable.SetValue(IsTrackedProperty, isTracked);
    }

    private static void OnIsTrackedChanged(BindableObject bindable, object oldValue, object newValue)
    {
        if ((bool)newValue)
        {
            var weakReference = new WeakReference(bindable);

            weakReferences.Add(weakReference);
        }
    }
}
```

I am using this class to track visual components in XAML like this:

```xml
<Grid local:MemoryTracker.IsTracked="True">
    <Label Text="{Binding}" />
</Grid>
```

I have prepared a small sample project that demonstrates the memory leak.

**Note:** It is possible that the following bug is related: [Memory leaks EVERYWHERE](https://github.com/dotnet/maui/issues/12039). Since I am describing a somewhat different scenario, I have decided to log a separate bug report for it.

### Steps to Reproduce

1. Build and run the sample project in *Release* build configuration.
2. Click on the "*Refresh Memory Info*" button to show the initial memory statistics.
3. Click on the "*Refresh Items Source*" button several times to recreate the collection of items.
4. Click on the "*Refresh Memory Info*" button to update the memory statistics.
5. Repeat the steps 3 and 4 several times to observe how both total tracked objects and alive objects constantly grow.
6. Click the *Recreate Control* button to recreate the `CollectionView` in the visual tree.
7. Click on the "*Refresh Memory Info*" button to update the memory statistics.
8. Observe how the alive objects are decreased after the `CollectionView` is recreated.

**Note:** Clicking the "*Refresh Memory Info*" button calls the garbage collector forcefully before updating the memory statistics:

```csharp
GC.Collect();
GC.WaitForPendingFinalizers();
GC.Collect();
```

### Link to public reproduction project repository

https://github.com/ivan-todorov-progress/maui-collection-view-memory-leak-bug

### Version with bug

7.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Windows

### Affected platform versions

N/A

**Comment by jsuarezruiz** [comment]
Could be related with https://github.com/dotnet/maui/pull/13327

**Comment by jonathanpeppers** [comment]
Yes, I think some of these recent fixes might solve this one:
* https://github.com/dotnet/maui/pull/13260
* https://github.com/dotnet/maui/pull/13333
* https://github.com/dotnet/maui/pull/13327

But I will test this case with main next week.

**Comment by LanceMcCarthy** [comment]
Thanks for the fast response, Redth and Javier! When you're done reviewing, can you please come back here and let me know which of those items was the culprit? Much appreciated!

![ZoolanderBossGIF](https://user-images.githubusercontent.com/3520532/219718277-816f6491-4e9c-426a-914e-b0b772357b06.gif)

**Comment by minze-it** [comment]
Those leaks are a major issue and an absolute show stopper for our app causing the dotnet-maui version not going live. It would be so great to have those fixes in the next service release of the dotnet 7 version.

**Comment by jonathanpeppers** [comment]
@ivan-todorov-progress I did indeed find an issue with your sample app above: https://github.com/dotnet/maui/pull/13530

Do you know if this only happens on Windows? Android actually seems like it is OK without me changing anything:

![image](https://user-images.githubusercontent.com/840039/220982328-c28ba3d7-6354-4815-a52e-43b7d4851837.png)

**Comment by ivan-todorov-progress** [comment]
@jonathanpeppers I can confirm this is not happening on Android. My speculation is that on Android the CollectionView clears the BindingContext of its items when they are recycled, so the Bindings get disconnected and we have no memory leak. On Windows it is easily reproducible, however. I have not had the chance to test it with the latest version of Visual Studio and .NET MAUI to see whether the leak is fixed. I'll keep you posted with my findings, once I upgrade.

**Comment by ivan-todorov-progress** [comment]
@jonathanpeppers I have updated Visual Studio to the latest 17.5 version and re-tested the sample app on all platforms. On Android and iOS there is no memory leak. The memory leak on Windows is still there, though. I believe, we need your fix about ClearLogicalChildren as well: https://github.com/jonathanpeppers/maui/commit/c7837db524233420767c078428981b66d548c1d8.

**Comment by JanZeman** [comment]
Hello @jonathanpeppers, I know this issue is closed some weeks ago. Nevertheless, can you please shortly comment on my first question on [this PR](https://github.com/ivan-todorov-progress/maui-collection-view-memory-leak-bug/pull/1)? Does you change fixes ObservableCollection behavior as well?

**Comment by JanZeman** [comment]
Hi @rmarinho, I can see this fix got merged to release/7.0.2xx. Can you please shortly explain or give reference to the release model? Is there any description that will explain which Visual Studio version will bring this fix to my local environment so I can try to verify and experiment? Thank you very much in advance.
